pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -222,48 +195,58 @@ def _get_noiseless_hamiltonian(self, leakage: bool) -> Hamiltonian:
             )
         else:
             noise = NoiseModel()
+
+        noiseless_data = HamiltonianData(
+            self.samples_obj, self._register, self._device, noise
+        )
         return Hamiltonian(
-            self.samples_obj,
-            self._register,
-            self._hamiltonian.data._device,
+            noiseless_data.samples,
+            noiseless_data.noise_trajectories[0][0],
+            noiseless_data.basis_data,
+            noiseless_data.lindblad_data,
             self._sampling_rate,
-            noise,
         )
 
     @property
-    def sampling_times(self) -> np.ndarray:
-        """The times at which hamiltonian is sampled."""
-        return self._hamiltonian.sampling_times
+    def _hamiltonians(self) -> Iterator[tuple[Hamiltonian, int]]:
+        for traj, noisy_samples, n in self._hamiltonian_data.noisy_samples:
+            yield Hamiltonian(
+                noisy_samples,
+                traj,
+                self._hamiltonian_data.basis_data,
+                self._hamiltonian_data.lindblad_data,
+                self._sampling_rate,
+            ), n
```

Comment:
Again, I think a NamedTuple for stuff like this would greatly improve code readability. If not that, then at least let's give a better name than `n` to the variable